### DISCUSSION TITLE
Handling the manufacturer id separately from the manufacturer data

### DISCUSSION THREAD

**Discussion by 1technophile** [theengs/decoder#100] [General]
Currently the manufacturerdata string contains the manufacturer id, for example:
```json
{"id":"F3","mac_type":1,"name":"F3","manufacturerdata":"330117540e101776666663f801370100","rssi":-85,"brand":"BlueMaestro","model":"TempoDisc","model_id":"BM_V23","tempc":30.9,"tempf":87.62,"hum":101.6,"dp":31.1,"volt":2.56}
```
Where nRF Connect gives:
Company 0x0133
Data 0x17540e101776666663f801370100

This is due to the first usage of the library with ESP32 but can cause issues with other implementations.

Do you think we should fix this and differentiate these 2 fields in the decoder and impact the client projects like OMG and Theengs Gateway?

**Comment by DigiH** [comment]
@1technophile if you remember the long discussion with the initial H505 decoder, when nRF Connect and even I/we thought that the beginning on the manufacturerdata was the company ID, but it turned out to be the ending part of the MAC address?

Unless doing an online check for 'proper' registered company IDs only I don't think this is a viable, secure idea. For decoder, nor for nRF Connect ;) EDIT, even with the online lookup it doesn't work, as in the below case!

https://github.com/theengs/decoder/issues/92

nRF Connect screenshot halfway down - can't seem to get an anchor for it ;(

**Reply by 1technophile** [reply]
> @1technophile if you remember the long discussion with the initial H505 decoder, when nRF Connect and even I/we thought that the beginning on the manufacturerdata was the company ID, but it turned out to be the ending part of the MAC address?

I'm wondering if it is an edge case or something that we will encounter a lot.

@h2zero do you know why the esp32 implementation doesn't differentiate the company ID as nRF Connect do?

**Reply by DigiH** [reply]
> I'm wondering if it is an edge case or something that we will encounter a lot.

Looking at the existing 30 manufacturerdata test cases there are only 10 which clearly adhere to the Company Identifier convention
```
{"RuuviTag RAWv1", "RuuviTag", "990403291A1ECE1EFC18F94202CA0B53"},
{"RuuviTag RAWv1", "RuuviTag maximum values", "990403FF7F63FFFF7FFF7FFF7FFFFFFF"},
{"RuuviTag RAWv1", "RuuviTag minimum values", "99040300FF6300008001800180010000"},
{"RuuviTag RAWv2", "RuuviTag", "99040512FC5394C37C0004FFFC040CAC364200CDCBB8334C884F"},
{"RuuviTag RAWv2", "RuuviTag maximum values", "9904057FFFFFFEFFFE7FFF7FFF7FFFFFDEFEFFFECBB8334C884F"},
{"RuuviTag RAWv2", "RuuviTag minimum values", "9904058001000000008001800180010000000000CBB8334C884F"},
{"iBeacon", "BlueCharm_135727", "4c000215426c7565436861726d426561636f6e730efe1355c5"},
{"iBeacon", "NRF51822", "4c000215fda50693a4e24fb1afcfc6eb07647825000100021a"},
{"BM_V23", "V23", "330117560e10177000ef01b3006c0100"},
{"MS-CDP", "Windows 10 Desktop", "060001092002ac6d90ec0132b3204cd39c7ced3e48436ba15dc6314778"},
```

6 with more or less questionable company results for their brand
```
{"Inkbird TH2", "tps", "660a03150110805908"}, - JUSTMORPH PTE. LTD
{"Inkbird TH1", "sps", "660a03150110805908"}, - JUSTMORPH PTE. LTD
{"TPMS", "TPMS1_10CA8F", "000180eaca10ca8ff46503007c0c00003300"}, - TomTom International BV
{"WS02", "ThermoBeacon", "100000001a110000f770580cf5016c0443090000"}, - Mitel Semiconductor
{"WS08", "ThermoBeacon", "110000001a110000f770580cf5016c0443090000"}, - Widcomm, Inc
{"H5102", "GVH5102_1234", "0100010103590e64"}, - Nokia Mobile Phones
```

5 with no registered company identifier - would be interesting to see what nRF Connect displays here
```
{"iNode", "test1", "90826300f0cf0000c409a20080"},
{"iNode", "test2", "9082dd0061b80000c409a00080"},
{"Inkbird TH2", "tps", "76fb03150110805908"}, - similar/same Inkbird TH2 model as above, possibly another case of MAC address part!?!
{"H5075", "GVH5075_1234", "88ec000418ee6400"},
{"H5072", "GVH5072_1234", "88ec0004344b6400"},
```

and 9 all starting with '000', all being 'Ericsson Technology Licensing'?!?
```
{"IBT-2X", "iBBQ", "0000fc45c30c458ee600e600"},
{"IBT-2X", "iBBQ", "0000fc45c30c458e18014001"},
{"IBT-2X", "iBBQ", "0000fc45c30c458ef6ff8a02"},
{"IBT-2X", "iBBQ", "00000000fc45c30d38a8dc00d200"},
{"IBT-2X", "iBBQ", "00000000fc45c30d38a8f6ff4402"},
{"IBT-6XS", "iBBQ", "00003403de2745cdd200c800f6ffd200f6fff6ff"},
{"SOLIS_6", "iBBQ", "000000000cb2b71b5b18c800c800f6fff6fff6fff6ff"},
{"IBT-4XS", "iBBQ", "0000000010082c40abe604010401fa00fa00"},
{"IBT-4XS", "iBBQ", "0000000010082c40abe60401f6ff58021202"},
```

**CORRECTION**: For the Inkbird TH2 models 

```
"tempc":{
         "decoder":["value_from_hex_data", "manufacturerdata", 0, 4, true],
         "post_proc":["/", 100]
      },
```
it's actually the temperature - so

`{"Inkbird TH2", "tps", "660a03150110805908"}, - JUSTMORPH PTE. LTD`

which nRF Connect probably shows as company identifier, depending on the current temperature, and

`{"Inkbird TH2", "tps", "76fb03150110805908"},`

with no company ID both show that here it definitely isn't ;) same for Inkbird TH1.

I really think this is an issue which nRF Connect should address, as it provides clearly false and misleading information in many cases.

**Reply by h2zero** [reply]
The manufacturer id is the first 2 bytes of the manufacturer data. The could be easily extracted but I'm not sure on it's usefulness. The only time the decoder uses that data is to detect iBeacons (0x4C00 = Apple) but any device conforming to that manufacturers specs could also use this ID, which is very common, so only in certain cases could I see it being useful.

As @DigiH pointed to above, this ID isn't always present, especially for devices that come from manufacturers without a registered ID ( most stuff from AliExpress).

**Reply by DigiH** [reply]
I think the issue is that the first 2 bytes of the manufacturerdata being the company identifier is **NOT** mandatory, leaving the manufacturerdata open to many individual uses, as above with it indicating the temperature, or with the Govee H5005 using it for parts of the MAC address.

That's what I meant with nRF Connect always presenting it as the company identifier being somehow a problematic issue, which I don't think Decoder should copy.

**Reply by h2zero** [reply]
> I think the issue is that the first 2 bytes of the manufacturerdata being the company identifier is NOT mandatory

Actually it is mandatory, that's why nRFconnect displays it that way.

Here's what the spec says:

> 1.4.1 Description The Manufacturer Specific data type is used for manufacturer specific data. 
**The first two data octets shall contain a company identifier from Assigned Numbers**. The interpretation of any other octets within the data shall be defined by the manufacturer specified by the company identifier.

Unfortunately there are a lot of non-conforming devices out there because they don't want to pay for their own ID and try to keep their specs closed source.

**Reply by DigiH** [reply]
> Actually it is mandatory, that's why nRF Connect displays it that way.

Ouch :( then that's not good that quite a few don't follow this. Not that I have any of their devices, but from a first look at their sites and devices I wouldn't have taken Inkbird or Govee for 'cheap' AliExpress-like devices ;)

For real world usage though this means that either only allow 'conforming' devices, or state that displaying the company identifier is not always a certain thing.

**Reply by h2zero** [reply]
What it does say about those devices is that they have not undergone the specifications testing and would fail if they did. This means that they cannot use the Bluetooth trademarks on their products. 

They get away with it though because they are operating within their own ecosystem where their apps etc ignore this fault. Only people like us who are collecting data from these without knowing their specs would notice this.

**Reply by emericg** [reply]
The list of manufacturer is not easily extensible, I'm sure it's harder and takes more time to get into that official list than to develop an entire BLE device from scratch...
But the manufacturer field is mandatory to get a valid packet, even if the value is 0 or just garbage...
It's easy enough to bypass that problem and just use service data instead of manufacturer data though.

I'm using a desktop Bluetooth stack (all three BLE stacks actually), and the manufacturer data does not contain the manufacturer ID, which is usually an int16 you can get separately. That make it weird to re-make a big endian hexadecimal string (even saying it is weird :) representation of the manufacturer ID and put it before the actual data for theengs.

**Reply by h2zero** [reply]
Yes, it's up to the application programmer to use the correct ID in the manufacturer data. Not a problem for testing things or custom applications but commercial products should adhere to the specs, when they don't, we get conversations such as this lol.

Anyway, we should just ignore this data IMO.

**Comment by DigiH** [comment]
Also a bit confused now with situations like

https://community.openmqttgateway.com/t/is-it-possible-to-please-add-inkbird-ibt-6x-to-compatibility-list-for-openmqttgateway/1850

as a similar thing happened with the IBT_2X.

@1technophile do you have an IBT_2X and implemented the original decoder yourself with its leading 4 "0000" at the beginning of the manufacturerdata, or was the original 4 "0000" manufacturerdata user reported?

In the latter case it could be easily missed that the 'suposed' company identifier was actually another four "0000" before, making it a total of 8 "00000000" as is this current case with the IBT_6X(S).

So maybe there are only 8 "00000000" versions, but depending on where the manufacturerdata reports come from it can easily be mistaken for just 4 "0000".

Or there really are two different IBT_2X versions and IBT_6XS and IBT_6X differ there.

Just asking before I create similar conditions and underscore variants for the IBT_6XS decoder.

**Comment by 1technophile** [comment]
> @1technophile do you have an IBT_2X and implemented the original decoder yourself with its leading 4 "0000" at the beginning of the manufacturerdata, or was the original 4 "0000" manufacturerdata user reported?

It was reported by a user [here](https://github.com/1technophile/OpenMQTTGateway/issues/1083) with the 0000
And a different [here ](https://community.home-assistant.io/t/presence-detection-ble-mi-flora-rf-gateway-with-esp32-openmqttgateway-is-now-compatible/37577/103) with 00000000
So 2 different users, 2 different types of advertisements.... Seems that we need to handle both also for the 6XS

I have only an IBT-4XS.

**Reply by DigiH** [reply]
> It was reported by a user [here](https://github.com/1technophile/OpenMQTTGateway/issues/1083) with the 0000

Yes, exactly what I thought, even that should have been "00000000", as the extra 4 "0000" reported as company identifier do actually belong to the beginning of the manufacturerdate. The same as with the current forum query, which threw me initially as well.

So there are only ever  8 "00000000" versions, and I'm now even thinking that the Solis 6 is just a rebranded IBT_6X(S)

```
000000000cb2b71b5b18c800c800f6fff6fff6fff6ff      SOLIS
00000000780473ee3984e6004a01dc00dc00e600f6ff      IBT-6X
    00003403de2745cdd200c800f6ffd200f6fff6ff      IBT-6XS - wrong without the leading 0000
```

where the last case is the wrongly implemented and test case 4 "0000" version, which doesn't really exist and is only due to wrong manufacturerdate reporting through nRF Connect screenshots.

**Reply by DigiH** [reply]
I think we can safely delete all 4 "0000"/8 "00000000" variations for IBT_2X and even pull together the Solis 6 probe and the IBT_6X(S), only using 8 "00000000" for them all - any indication where the initial IBT_6XS manufacturerdata came from?

**Reply by DigiH** [reply]
Found it

https://community.openmqttgateway.com/t/ble-wireless-bbq-temperature-probe-with-esp32/1561/14

Same thing there, the "0000" company identifier should be included at the beginning of the manufacturerdata.

I think the current IBT_6XS decoder doesn't actually decode any IBT_6XS device because of the `"manufacturerdata","=", 40,` restriction, which should be `"manufacturerdata","=", 44,`

**Reply by 1technophile** [reply]
> I think the current IBT_6XS decoder doesn't actually decode any IBT_6XS device because of the "manufacturerdata","=", 40, restriction, which should be "manufacturerdata","=", 44,

I agree

**Reply by DigiH** [reply]
We'll have to make all these changes for the IBT_2X and IBT_6XS and pulling it together with the Solis 6, including converting all the test cases to "00000000" only cases. We just need to decide on naming the IBT_6X(s)/Solis brand, model and model_id in a unified decoder.

I'm happy to have a go at it tomorrow.

**Reply by 1technophile** [reply]
Maybe we just put the 2 models in the doc (.md) file and let the decoder with Inkbird brand and model. Thoughts?

**Reply by h2zero** [reply]
Cool! Why not just make the brand `Inkbird/Solis`? Then we can skip the docs and have minimal impact on size.

**Reply by 1technophile** [reply]
I'm ok with this. Just wanted to reduce the payload size for esp.

**Reply by h2zero** [reply]
12 bytes is hopefully not a concern 😄, there are bigger fish to fry in that regard.

**Reply by DigiH** [reply]
I left the docs separately, as they are uniquely different hardware, both linking to the same merged decoder.

**Comment by DigiH** [comment]
Picking this up again with the new TheengsExplorer also showing the company names, as many other BLE scanning apps do, and my comment in the initial Explorer discussion.

https://github.com/theengs/decoder/discussions/132#discussioncomment-3048555

Since we pretty much know if a device adheres to the company ID convention or not when creating decoders. How about adding a simple static boolean flag to each decoder, something like "cidc" for "company ID compliant", so that every app using the Decoder library can easily decide to display a proper company name or not. This would be short from actually doing a company name lookup in Decoder itself, which would also require to keep the list updated. 

Or adding
```
"mfid":{
         "decoder":["string_from_hex_data", "manufacturerdata", 0, 4]
      },
```
to all decoders, with non compliant ones having an empty or "FFFF" value.

Just an idea to get to grips with this issue in some way.

**Reply by koenvervloesem** [reply]
Good idea to try to distinguisgh these non-compliant devices, so tools such as Theengs Explorer can take this into account.

As non-compliant devices are the exception, I would just add a flag for those devices that are known to use the company ID for their values, to not overcomplicate all other decoders, no?

**Reply by DigiH** [reply]
> As non-compliant devices are the exception

Not quite unfortunately ;) so I think it makes it even more important to decide on a solution in Decoder, so that any app using it can make confident decisions about displaying the manufacturer ID or not.

**Reply by DigiH** [reply]
So either

```
"mfid":{
         "decoder":["static_value", "FFFF"]
      },
```

to align with the official
`
This value is reserved as the default vendor ID when no Device ID service record is present in a remote device.
`

Leaving all compliant devices not sending a **mfid**, apart from iBeacons. Which again would make a double check necessary for any Decoder using app, for **mfid** AND IS "FFFF, while a separate

```
"cidc":{
         "decoder":["static_value", false]
      },
```

could allow only checking for the **cidc** existence, when only added to non-cidc decoders.

**Reply by DigiH** [reply]
Did a sample branch, also for testing and eventual implementation with Explorer, with a few obvious (cid is actual temperature) and non-registered sample decoders

https://github.com/DigiH/decoder/tree/cidc-introduction
